File: Advent of Code 2015/Day22.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Advent_of_Code_2015
{
    /// <summary>
    /// Day 22: Wizard Simulator 20XX
    /// Turn-based RPG combat as a wizard with spells.
    /// Player has 50 HP, 500 mana. Boss stats from input.
    /// Find minimum mana to win (Part 1) and minimum mana on Hard mode (Part 2).
    /// </summary>
    [TestClass]
    public class Day22
    {
        private static readonly string inputString = InputLoader.ReadAllText("day22.txt");

        record Spell(string Name, int Cost, int Damage, int Heal, int Armor, int Mana, int Duration);
        record State(int PlayerHP, int PlayerMana, int BossHP, int ShieldTimer, int PoisonTimer, int RechargeTimer, int ManaSpent, bool PlayerTurn);

        static readonly List<Spell> Spells = new()
        {
            new("Magic Missile", 53, 4, 0, 0, 0, 0),
            new("Drain", 73, 2, 2, 0, 0, 0),
            new("Shield", 113, 0, 0, 7, 0, 6),
            new("Poison", 173, 0, 0, 0, 0, 6),
            new("Recharge", 229, 0, 0, 0, 101, 5)
        };

        static (int hp, int damage) ParseBoss(string input)
        {
            var lines = input.Split('\n', StringSplitOptions.RemoveEmptyEntries);
            int hp = int.Parse(lines[0].Split(':')[1].Trim());
            int damage = int.Parse(lines[1].Split(':')[1].Trim());
            return (hp, damage);
        }

        static int FindMinMana(int bossHP, int bossDamage, bool hardMode)
        {
            var initial = new State(50, 500, bossHP, 0, 0, 0, 0, true);
            var queue = new PriorityQueue<State, int>();
            queue.Enqueue(initial, 0);
            var visited = new HashSet<(int, int, int, int, int, int, bool)>();
            int minMana = int.MaxValue;

            while (queue.Count > 0)
            {
                var state = queue.Dequeue();

                if (state.ManaSpent >= minMana) continue;

                var key = (state.PlayerHP, state.PlayerMana, state.BossHP, state.ShieldTimer, state.PoisonTimer, state.RechargeTimer, state.PlayerTurn);
                if (visited.Contains(key)) continue;
                visited.Add(key);

                var (hp, mana, bossHp, shield, poison, recharge, spent, playerTurn) = state;

                // Hard mode: lose 1 HP at start of player turn
                if (hardMode && playerTurn)
                {
                    hp--;
                    if (hp <= 0) continue;
                }

                // Apply effects
                int armor = 0;
                if (shield > 0)
                {
                    armor = 7;
                    shield--;
                }
                if (poison > 0)
                {
                    bossHp -= 3;
                    poison--;
                }
                if (recharge > 0)
                {
                    mana += 101;
                    recharge--;
                }

                // Check if boss is dead
                if (bossHp <= 0)
                {
                    minMana = Math.Min(minMana, spent);
                    continue;
                }

                if (playerTurn)
                {
                    // Try each spell
                    foreach (var spell in Spells)
                    {
                        if (mana < spell.Cost) continue;

                        // Check if effect is already active
                        if (spell.Name == "Shield" && shield > 0) continue;
                        if (spell.Name == "Poison" && poison > 0) continue;
                        if (spell.Name == "Recharge" && recharge > 0) continue;

                        int newHP = hp + spell.Heal;
                        int newMana = mana - spell.Cost;
                        int newBossHP = bossHp - spell.Damage;
                        int newShield = shield + (spell.Name == "Shield" ? spell.Duration : 0);
                        int newPoison = poison + (spell.Name == "Poison" ? spell.Duration : 0);
                        int newRecharge = recharge + (spell.Name == "Recharge" ? spell.Duration : 0);
                        int newSpent = spent + spell.Cost;

                        var newState = new State(newHP, newMana, newBossHP, newShield, newPoison, newRecharge, newSpent, false);
                        queue.Enqueue(newState, newSpent);
                    }
                }
                else
                {
                    // Boss turn
                    int damage = Math.Max(1, bossDamage - armor);
                    int newHP = hp - damage;
                    if (newHP > 0)
                    {
                        var newState = new State(newHP, mana, bossHp, shield, poison, recharge, spent, true);
                        queue.Enqueue(newState, spent);
                    }
                }
            }

            return minMana;
        }

        [TestMethod]
        public void Part1Examples()
        {
            // Example 1: Player has 10 HP, 250 mana, Boss has 13 HP, 8 damage
            int result1 = FindMinMana(13, 8, false);
            Console.WriteLine($"Example 1: {result1} mana");
            // Expected: Can win with Poison (173) then Magic Missile (53) or other combos

            // Example 2: Player has 10 HP, 250 mana, Boss has 14 HP, 8 damage
            int result2 = FindMinMana(14, 8, false);
            Console.WriteLine($"Example 2: {result2} mana");
        }

        [TestMethod]
        public void Part1()
        {
            var (bossHP, bossDamage) = ParseBoss(inputString);
            int minMana = FindMinMana(bossHP, bossDamage, false);
            Assert.AreEqual(900, minMana);
            Console.WriteLine($"Part 1: Minimum mana to win = {minMana}");
        }

        [TestMethod]
        public void Part2Examples()
        {
            // Hard mode: lose 1 HP at start of each player turn
            var (bossHP, bossDamage) = ParseBoss(inputString);
            int minMana = FindMinMana(bossHP, bossDamage, true);
            Console.WriteLine($"Part 2 test: {minMana}");
        }

        [TestMethod]
        public void Part2()
        {
            var (bossHP, bossDamage) = ParseBoss(inputString);
            int minMana = FindMinMana(bossHP, bossDamage, true);
            Assert.AreEqual(1216, minMana);
            Console.WriteLine($"Part 2: Minimum mana on Hard mode = {minMana}");
        }
    }
}

File: rust/src/day22.rs
use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashSet};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
struct State {
    player_hp: i32,
    player_mana: i32,
    boss_hp: i32,
    boss_damage: i32,
    shield: i32,
    poison: i32,
    recharge: i32,
    spent: i32,
    player_turn: bool,
}

#[derive(Clone, Copy)]
struct Spell {
    cost: i32,
    damage: i32,
    heal: i32,
    armor: i32,
    mana: i32,
    duration: i32,
    kind: SpellKind,
}

#[derive(Clone, Copy, PartialEq, Eq)]
enum SpellKind {
    Instant,
    Shield,
    Poison,
    Recharge,
}

const SPELLS: [Spell; 5] = [
    Spell { cost: 53, damage: 4, heal: 0, armor: 0, mana: 0, duration: 0, kind: SpellKind::Instant },
    Spell { cost: 73, damage: 2, heal: 2, armor: 0, mana: 0, duration: 0, kind: SpellKind::Instant },
    Spell { cost: 113, damage: 0, heal: 0, armor: 7, mana: 0, duration: 6, kind: SpellKind::Shield },
    Spell { cost: 173, damage: 0, heal: 0, armor: 0, mana: 0, duration: 6, kind: SpellKind::Poison },
    Spell { cost: 229, damage: 0, heal: 0, armor: 0, mana: 101, duration: 5, kind: SpellKind::Recharge },
];

fn parse_boss(input: &str) -> (i32, i32) {
    let mut hp = 0;
    let mut damage = 0;
    for line in input.lines() {
        let mut parts = line.split(':');
        let key = parts.next().unwrap().trim();
        let val: i32 = parts.next().unwrap().trim().parse().unwrap();
        match key {
            "Hit Points" => hp = val,
            "Damage" => damage = val,
            _ => {}
        }
    }
    (hp, damage)
}

fn apply_effects(state: &mut State) {
    if state.shield > 0 {
        state.shield -= 1;
    }
    if state.poison > 0 {
        state.boss_hp -= 3;
        state.poison -= 1;
    }
    if state.recharge > 0 {
        state.player_mana += 101;
        state.recharge -= 1;
    }
}

fn min_mana_to_win(boss_hp: i32, boss_damage: i32, hard: bool) -> i32 {
    let start = State {
        player_hp: 50,
        player_mana: 500,
        boss_hp,
        boss_damage,
        shield: 0,
        poison: 0,
        recharge: 0,
        spent: 0,
        player_turn: true,
    };

    let mut heap = BinaryHeap::new();
    heap.push((Reverse(0), start));
    let mut seen: HashSet<State> = HashSet::new();
    let mut best = i32::MAX;

    while let Some((Reverse(cost), mut s)) = heap.pop() {
        if cost >= best {
            continue;
        }
        if !seen.insert(s) {
            continue;
        }

        if hard && s.player_turn {
            s.player_hp -= 1;
            if s.player_hp <= 0 {
                continue;
            }
        }

        apply_effects(&mut s);
        if s.boss_hp <= 0 {
            best = best.min(cost);
            continue;
        }

        if s.player_turn {
            for spell in SPELLS.iter() {
                if s.player_mana < spell.cost {
                    continue;
                }
                // prevent re-casting active effects
                if spell.kind == SpellKind::Shield && s.shield > 0 {
                    continue;
                }
                if spell.kind == SpellKind::Poison && s.poison > 0 {
                    continue;
                }
                if spell.kind == SpellKind::Recharge && s.recharge > 0 {
                    continue;
                }

                let mut ns = s;
                ns.player_turn = false;
                ns.player_mana -= spell.cost;
                ns.spent += spell.cost;
                ns.boss_hp -= spell.damage;
                ns.player_hp += spell.heal;
                match spell.kind {
                    SpellKind::Shield => ns.shield = spell.duration,
                    SpellKind::Poison => ns.poison = spell.duration,
                    SpellKind::Recharge => ns.recharge = spell.duration,
                    SpellKind::Instant => {}
                }

                if ns.boss_hp <= 0 {
                    best = best.min(ns.spent);
                } else {
                    heap.push((Reverse(ns.spent), ns));
                }
            }
        } else {
            let armor = if s.shield > 0 { 7 } else { 0 };
            let dmg = (s.boss_damage - armor).max(1);
            s.player_hp -= dmg;
            if s.player_hp > 0 {
                s.player_turn = true;
                heap.push((Reverse(s.spent), s));
            }
        }
    }
    best
}

pub fn part1(input: &str) -> i32 {
    let (hp, dmg) = parse_boss(input);
    min_mana_to_win(hp, dmg, false)
}

pub fn part2(input: &str) -> i32 {
    let (hp, dmg) = parse_boss(input);
    min_mana_to_win(hp, dmg, true)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::read_input;

    #[test]
    fn answers() {
        let input = read_input(22).expect("input");
        assert_eq!(900, part1(&input));
        assert_eq!(1216, part2(&input));
    }
}
